src/google_sheets.rs

```diff
@@ -0,0 +1,247 @@
+extern crate yup_oauth2 as oauth;
+
+use std::error;
+use std::fmt;
+
+use oauth::{AccessToken, InstalledFlowAuthenticator, InstalledFlowReturnMethod};
+use reqwest::{header, Client, Method, Request, StatusCode, Url};
+use serde::{Deserialize, Serialize};
+
+/// Endpoint for the Google Sheets API.
+const BASE_ENDPOINT: &str = "https://sheets.googleapis.com/v4/";
+
+pub trait IntoSheetEntry {
+    fn into_sheet_entry(&self) -> Vec<String>;
+}
+
+#[derive(Serialize)]
+pub struct EmptyBody {}
+
+pub struct Sheets {
+    token: AccessToken,
+    client: Client,
+    sheet_id: String,
+}
+
+impl Sheets {
+    pub fn get_link_to_sheet(sheet_id: &str) -> String {
+        format!("https://docs.google.com/spreadsheets/d/{}/", sheet_id)
+    }
+    pub async fn initialize(sheet_id: &str) -> Result<Self, APIError> {
+        let token = match Sheets::authenticate().await {
+            Ok(t) => t,
+            Err(e) => return Err(e),
+        };
+
+        Sheets::new(token, sheet_id)
+    }
+
+    pub async fn authenticate() -> Result<AccessToken, APIError> {
+        // Read application secret from a file. Sometimes it's easier to compile it directly into
+        // the binary. The `client_secret` file contains JSON like `{"installed":{"client_id": ... }}`
+        let secret = oauth::read_application_secret("client_secret.json")
+            .await
+            .expect("client_secret.json");
+
+        // All authentication tokens are persisted to a file named `tokencache.json`.
+        // The authenticator takes care of caching tokens to disk and refreshing tokens once they've expired.
+        let auth = match InstalledFlowAuthenticator::builder(
+            secret,
+            InstalledFlowReturnMethod::HTTPRedirect,
+        )
+        .persist_tokens_to_disk("tokencache.json")
+        .build()
+        .await
+        {
+            Ok(a) => a,
+            Err(e) => return Err(APIError::from(e)),
+        };
+
+        let scope = &["https://www.googleapis.com/auth/spreadsheets"];
+        match auth.token(scope).await {
+            Ok(tok) => Ok(tok),
+            Err(err) => panic!("Could not authenticate properly {:?}", err),
+        }
+    }
+
+    pub fn new(token: AccessToken, sheet_id: &str) -> Result<Self, APIError> {
+        match Client::builder().build() {
+            Ok(client) => Ok(Self {
+                token,
+                client,
+                sheet_id: String::from(sheet_id),
+            }),
+            Err(_) => Err(APIError {
+                status_code: StatusCode::from_u16(500).unwrap(),
+                body: "Could not instantiate client".to_string(),
+            }),
+        }
+    }
+
+    pub fn request<T>(
```

Comment:
Yes, I should go through and add doc comments everywhere.

This is not exactly for end-users, but I guess it could be? I think I should remove the `pub` for now, though.